packages/cursorless-engine/src/core/commandVersionUpgrades/generateSpokenForm.ts

```diff
@@ -0,0 +1,478 @@
+        const connective = getRangeConnective(
+          modifier.excludeAnchor,
+          modifier.excludeActive,
+        );
+        return [anchor, connective, active];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+    }
+
+    default:
+      return [modifiers[modifier.type]];
+  }
+}
+
+function relativeScopeInclusiveToSpokenForm(
+  modifier: RelativeScopeModifier,
+): RecursiveArray<string> {
+  const scope = scopeTypeToSpokenForm(modifier.scopeType);
+
+  if (modifier.length === 1) {
+    const direction =
+      modifier.direction === "forward"
+        ? connectives.forward
+        : connectives.backward;
+
+    // token forward/backward
+    return [scope, direction];
+  }
+
+  const length = numberToSpokenForm(modifier.length);
+  const scopePlural = pluralize(scope);
+
+  // two tokens
+  // This could also have been "two tokens forward"; there is no way to disambiguate.
+  if (modifier.direction === "forward") {
+    return [length, scopePlural];
+  }
+
+  // two tokens backward
+  return [length, scopePlural, connectives.backward];
+}
+
+function relativeScopeExclusiveToSpokenForm(
+  modifier: RelativeScopeModifier,
+): RecursiveArray<string> {
+  const scope = scopeTypeToSpokenForm(modifier.scopeType);
+  const direction =
+    modifier.direction === "forward" ? connectives.next : connectives.previous;
+
+  if (modifier.offset === 1) {
+    const number = numberToSpokenForm(modifier.length);
+
+    if (modifier.length === 1) {
+      // next/previous token
+      return [direction, scope];
+    }
+
+    const scopePlural = pluralize(scope);
+
+    // next/previous two tokens
+    return [direction, number, scopePlural];
+  }
+
+  if (modifier.length === 1) {
+    const ordinal = ordinalToSpokenForm(modifier.offset);
+    // second next/previous token
+    return [ordinal, direction, scope];
+  }
+
+  throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+}
+
+function scopeTypeToSpokenForm(scopeType: ScopeType): string {
+  switch (scopeType.type) {
+    case "oneOf":
+    case "customRegex":
+    case "switchStatementSubject":
+      throw new NoSpokenFormError(`Scope type '${scopeType.type}'`);
+    case "string":
+      throw new AmbiguousSpokenFormError(`Scope type '${scopeType.type}'`);
+    case "surroundingPair": {
+      const pair = surroundingPairNameToSpokenForm(scopeType.delimiter);
+      if (scopeType.forceDirection != null) {
+        const direction =
+          scopeType.forceDirection === "left"
+            ? surroundingPairForceDirections.left
+            : surroundingPairForceDirections.right;
+        return `${direction} ${pair}`;
+      }
+      return pair;
+    }
+
+    default:
+      return scopeSpokenForms[scopeType.type];
+  }
+}
+
+function markToSpokenForm(mark: PartialMark): RecursiveArray<string> {
+  switch (mark.type) {
+    case "decoratedSymbol": {
+      const [color, shape] = mark.symbolColor.split("-");
```

Comment:
This is VSCode-specific. I'm inclined to get this stuff merged in, but we should definitely revisit the decorated symbol aspect of all this. My understanding was that the decorated symbols were to be editor-specific, so really VSCode should be injecting a function to convert decorated symbol to spoken form. But then obv we can't use this in CLI. Either way, let's leave it and file an issue to revisit it. Prob can do that when we get further along with other editors

Reply:
See https://github.com/cursorless-dev/cursorless/issues/1709